returnn/forward.py

```diff
@@ -0,0 +1,213 @@
+__all__ = ["ReturnnForwardJob"]
+
+from sisyphus import *
+
+import copy
+import glob
+import os
+import shutil
+import subprocess as sp
+import tempfile
+
+from i6_core.returnn.config import ReturnnConfig
+from i6_core.returnn.training import Checkpoint
+import i6_core.util as util
+
+Path = setup_path(__package__)
+
+
+class ReturnnForwardJob(Job):
+    """
+    Run a RETURNN "forward" pass to HDF with a specified model checkpoint.
+    Also allows to run an "eval" task pass, which is similar to "forward" but treats all layers as in training mode,
+    which can be used to e.g. do cheating experiments.
+
+    Outputs:
+
+    Dict[tk.Path] out_hdf_files: Dictionary of all output HDF files that were requested by the key list `hdf_outputs`
+    tk.Path out_default_hdf: For forward (not eval) mode, this contains the default HDF file that is always written
+        by RETURNN, independent of other settings.
+    """
+
+    def __init__(
+        self,
+        model_checkpoint,
+        returnn_config,
+        hdf_outputs=None,
+        eval_mode=False,
+        *,  # args below are keyword only
+        log_verbosity=3,
+        device="gpu",
+        time_rqmt=4,
+        mem_rqmt=4,
+        cpu_rqmt=4,
+        returnn_python_exe=None,
+        returnn_root=None,
+    ):
+        """
+
+        :param Checkpoint model_checkpoint: Checkpoint object pointing to a stored RETURNN Tensorflow model
+        :param ReturnnConfig returnn_config: RETURNN config dict
+        :param dict returnn_post_config: RETURNN config dict (no hashing)
+        :param list[str] hdf_outputs: list of additional hdf output layer file names that the network generates (e.g. attention.hdf);
+          The hdf outputs have to be a valid subset or be equal to the hdf_dump_layers in the config.
+        :param bool eval_mode: run forward in eval mode, the default hdf is not available in this case and no search will be done.
+        :param int log_verbosity: RETURNN log verbosity
+        :param str device: RETURNN device, cpu or gpu
+        :param int time_rqmt: job time requirement
+        :param int mem_rqmt: job memory requirement
+        :param int cpu_rqmt: job cpu requirement
+        :param Path|str returnn_python_exe: path to the RETURNN executable (python binary or launch script)
+        :param Path|str returnn_root: path to the RETURNN src folder
+        """
+        self.returnn_python_exe = (
+            returnn_python_exe
+            if returnn_python_exe is not None
+            else gs.RETURNN_PYTHON_EXE
+        )
+        self.returnn_root = (
+            returnn_root if returnn_root is not None else gs.RETURNN_ROOT
+        )
+
+        self._model_checkpoint = model_checkpoint
+        self._returnn_config = returnn_config
+        self._eval_mode = eval_mode
+        self._log_verbosity = log_verbosity
+        self._device = device
```

Comment:
why do all these guys start with an underscore?

Reply:
I thought we could start making rather private members private, but I guess this is somewhat inconsistent then.